File: validate.py
#!/usr/bin/env python

# Quick little program to validate the checksums in the output files of
# new_test and daemon
#
# Note: The C++ version is almost 20X faster.  Use it unless there's a very
# good reason not to.
#
# Usage:  validate.py <file1> [ <file2> ... ]
#


import sys
import struct

BLOCK_SIZE=1024

def main():
    if len(sys.argv) < 2:
        print "Usage:", sys.argv[0], "<file1> [ <file2> ... ]"
        return
    
    print "Input files:",
    for fname in sys.argv[1:]:
        print fname,
    print
    
    for fname in sys.argv[1:]:
        validate_file( fname)
                
def validate_file( fname):
    print fname, ":",
    f = open(fname, 'rb')
    checksum = 0;
    address = 0
    while True:
        val_str = f.read(BLOCK_SIZE)
        if len(val_str) == 0:
            print "validated"
            return  # normal exit
        elif len(val_str) < BLOCK_SIZE:
            print "Skipping last", len(val_str), "bytes."
            return
        
        checksum = 0
        format_str = "%dI" % (BLOCK_SIZE/4)
        vals = struct.unpack_from(format_str, val_str)
        if (len(vals) != BLOCK_SIZE/4):
            print "Unexpected block len: ", len(vals)

        for n in vals[:-1]:
            checksum += n
        
        checksum = checksum % 0x100000000
        if checksum != vals[-1]:
            print ": Checksum error for block at 0x%x (0x%x)" % (address, address+BLOCK_SIZE-4)
            print "Calculated checksum: %d   File's checksum: %d" %(checksum, vals[-1])
            return  
#            else:
#                print "checksum pass for block at", address
        address += BLOCK_SIZE




if __name__ == "__main__":
    main()

File: validate.cpp
// A quick little program that will validate the checksums new_test and
// daemon write into their data files.
//
// NOTE: The algorithm this code uses must obviously match the algorithm
// that the client uses!  See initBuffer() in new_test.cpp!

#include <stdint.h>

#include <fstream>
#include <iostream>
#include <string>
using namespace std;


void validateFile( const string &fname)
{

    ifstream infile(fname.c_str(), ios_base::binary | ios_base::ate);
    streampos fileSize = infile.tellg();
    infile.seekg( 0);
    
    if (!infile) {
        cerr << "Failed to open " << fname << ".  Continuing with next file." << endl;
        return;
    } else {
        cout << fname << ": " << flush;
    }
    
    
    uint32_t vals[256];
    uint32_t checksum = 0;
    uint64_t address = 0;
    
    while (true) {
        streampos startPos = infile.tellg();
        infile.read( (char *)vals, 1024);
               
        if (infile.eof()) {
            if (fileSize != startPos) {
                cout << "Skipping last " << fileSize - startPos << " bytes. " << endl;
            } else {
                cout << "validated" << endl;
            }
            return;
        }
        
        checksum = 0;
        for (unsigned i = 0; i < 255; i++) {
            checksum += vals[i];
        }
        if (vals[255] != checksum) {
            cout << "Checksum error at 0x" << hex << address
                    << ".  Calculated checksum: " << dec << checksum
                    << "   File's checksum: " << vals[255] << endl;
            return;
        }
        
        address += 1024;
    }
    
    cerr << "BUG! This line should never be printed!" << endl;
    return;
}
 
 
 
int main( int argc, char **argv)
{
    if (argc < 2) {
        cerr << "Usage: " << argv[0] << "<filename> [<filename> ...]" << endl;
        return 1;
    }
    
    for (int i=1; i < argc; i++) {
        validateFile( argv[i]);
    }
    
    return 0;
}